objects.js

```diff
@@ -2640,87 +2640,137 @@ SpriteMorph.prototype.reporterize = function (expressionString) {
-            ch = expression[idx];
-            idx += 1;
-            switch (ch) {
-            case ' ':
-                break;
-            case '(':
-                if (inputs[operator ? 1 : 0].length) {
-                    inputs[operator ? 1 : 0] = [
-                        inputs[operator ? 1 : 0],
-                        parseInfix(nested())
-                    ];
+    function createASTFromTokens(tokens) {
+        return precedence2(tokens);
+        function precedence0(tokens) {
+            var token = tokens.shift();
+            if (token.type === "lp") {
+                var exp = createASTFromTokens(tokens);
+                if (tokens.shift().type !== "rp") {
+                    throw "Mismatched parenthesis";
+                }
+                return exp;
+            } else if (token.type === "number") {
+                return +token.val;
+            } else if (token.type === "operator") {
+                if (token.val === "-" || token.val === "+") {
+                    return [token.val, 0, precedence0(tokens)];
                 } else {
-                    inputs[operator ? 1 : 0] = parseInfix(nested());
+                    throw "Invalid prefix operator";
                 }
-                break;
-            case '-':
-            case '+':
-            case '*':
-            case '/':
-            case '%':
-                if (!operator && !inputs[0].length) {
-                    inputs[0] += ch;
-                } else if (operator) {
-                    return parseInfix(
-                        expression.slice(idx),
-                        ch,
-                        [operator, already, format(inputs[1])]
-                    );
+            } else if (token.type === "ident") {
+                if (tokens.length > 0 && tokens[0].type === "lp") {
+                    var args = [];
+                    tokens.shift();
+                    while (tokens[0].type !== "rp") {
+                        args.push(createASTFromTokens(tokens));
+                        if (tokens[0].type === "rp") {
+                            break;
+                        }
+                        if (tokens[0].type !== "comma") {
+                            throw "No comma between arguments";
+                        }
+                    }
+                    tokens.shift();
+                    return [token.val].concat(args);
                 } else {
-                    operator = ch;
-                    already = format(inputs[0]);
+                    return token.val;
                 }
-                break;
-            default:
-                inputs[operator ? 1 : 0] += ch;
+            } else {
+                throw "Invalid expression";
             }
         }
-        if (operator) {
-            return [operator, already, format(inputs[1])];
+        function precedence1(tokens) {
+            var exp = precedence0(tokens);
+            if (tokens.length === 0) {
```

Comment:
OMG Github just lost my comment. :(

Anyway... You're right! I've been not writing enough JS lately....

I was suggesting `!tokens.length` then, because that's more common in Snap!'s code too. But I don't have a strong preference...

(In Snap! .length is used as a boolean about 60-70 times, and `=== 0` only 15, according to my basic grepping.)